File: List-1-Java/src/NumberCoder/UBaseCoder.java
package NumberCoder;

import java.util.ArrayList;
import java.util.List;

public class UBaseCoder implements NumberCoder {
    private final int base;

    public UBaseCoder(int base) {
        this.base = base;
    }

    @Override
    public List<Byte> codeNumber(long number) {
        boolean isNegative = false;
        List<Byte> numberInBase = new ArrayList<>();

        if (number < 0) {
            isNegative = true;
            number *= -1;
        }

        while (number > 0) {
            byte remainder = (byte) (number % base);
            number /= base;

            numberInBase.add(remainder);
        }

        // Add additional bit for sign
        numberInBase.add((byte) 0);

        if (isNegative) {
            convertToNeg(numberInBase);
        }

        return numberInBase;
    }

    @Override
    public long decodeNumber(List<Byte> codedNumber) {
        long accumulator = 0;
        int factor = 0;

        for (Byte frag : codedNumber) {
            accumulator += (long) (frag * Math.pow(base, factor));
            factor += 1;
        }

        if (codedNumber.get(codedNumber.size() - 1) >= base / 2) {
            System.out.println(codedNumber.get(codedNumber.size() - 1));
            accumulator -= (long) Math.pow(base, factor);
        }

        return accumulator;
    }

    @Override
    public byte getBase() {
        return (byte) base;
    }

    private void convertToNeg(List<Byte> number) {
        List<Byte> convertedNumber = new ArrayList<>();
        for (byte mag : number) {
            convertedNumber.add((byte) (base - 1 - mag));
        }

        int carry = 1;

        for (int i = 0; i < convertedNumber.size(); i++) {
            convertedNumber.set(i, (byte) (convertedNumber.get(i) + carry));
            if (convertedNumber.get(i) >= base) {
                convertedNumber.set(i, (byte) (convertedNumber.get(i) - base));
                carry = 1;
            } else {
                carry = 0;
            }
        }

        number.clear();
        number.addAll(convertedNumber);
    }
}

File: List-1/NthBaseCalculatorLib/src/UBaseCoder.cpp
//
// Created by adrianczubaty on 07/11/23.
//

#include "../headers/UBaseCoder.h"

UBaseCoder& UBaseCoder::operator =(const UBaseCoder &rhs) {
    if (this == &rhs) {
        return *this;
    }

    this->_base = rhs._base;
    return *this;
}

CodedNumber UBaseCoder::operator()(int64_t number) {
    auto isNegative = false;
    CodedNumber numberInBase;

    if (number < 0) {
        isNegative = true;
        number *= -1;
    }

    while (number > 0) {
        uint8_t remainder = number % _base;
        number /= _base;

        numberInBase.push_back(remainder);
    }

    // Add additional bit for sign
    numberInBase.push_back(0);

    if (isNegative) {
        convertToNeg(numberInBase);
    }

    return numberInBase;
}

int64_t UBaseCoder::operator()(const CodedNumber &codedNumber) {
    auto accumulator = int64_t { 0 };
    auto factor { 0 };

    for (auto & frag : codedNumber) {
        accumulator += frag * std::pow(_base, factor);
        factor += 1;
    }

    if (codedNumber[codedNumber.size() - 1] >= getBase() / 2) {
        std::cout << *codedNumber.end() << std::endl;
        accumulator -= std::pow(_base, factor);
    }

    return accumulator;
}

void UBaseCoder::convertToNeg(CodedNumber &number) {
    auto convertedNumber = std::vector(number.size(), static_cast<uint8_t>(_base - 1));
    auto carry = 1;

    for (auto i = 0; i < number.size(); i++) {
        convertedNumber[i] -= number[i];
    }

    for (auto & mag : convertedNumber) {
        mag += carry;
        if (mag >= _base) {
            mag -= _base;
            carry = 1;
        } else {
            carry = 0;
        }
    }

    number = convertedNumber;
}